main.rb

```diff
@@ -0,0 +1,172 @@
+module Enumerable
+  def my_each
+    pos = 0
+    loop do
+      break if pos >= length
+
+      yield self[pos]
+      pos += 1
+    end
+  end
+
+  def my_each_with_index
+    pos = 0
+    loop do
+      break if pos >= length
+
+      yield self[pos], pos
+      pos += 1
+    end
+  end
+
+  def my_select
+    return_arr = []
+    my_each do |element|
+      return_arr.push(element) if yield element
+    end
+    return_arr
+  end
+
+  def my_all
+    return_val = true
+    my_each do |element|
+      return_val = yield element
+    end
+    return_val
+  end
+
+  def my_any
+    return_val = false
+    my_each do |element|
+      return_val = yield element
+    end
+    return_val
+  end
+
+  def my_none
+    return_val = true
+    my_each do |element|
+      return_val = !(yield element)
+    end
+    return_val
+  end
+
+  def my_count(arg = nil)
+    count_val = 0
+    my_each do |element|
+      count_val += 1 if (block_given? && yield(element)) || (element == arg)
+    end
+    count_val
+  end
+
+  def my_map(proc_arg = nil)
+    return_arr = []
+    my_each_with_index do |element, index|
+      if proc_arg
+        return_arr.push(proc_arg.call(element))
+      elsif block_given?
+        return_arr.push(yield(element))
+      else
+        return_arr.push(index)
+      end
+    end
+    return_arr
+  end
```

Comment:
* [ ] `my_map` is identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_map returns an Enumerator if no block is given
   ```ruby
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_map.class # not an Enumerator
  ```